articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
+The `WITH` reduces the result set down to one row and then the final part will delete everything in the database.
+
+This attack was not possible to avoid using parameterization.
+To avoid this attack sanitization must be used.
+It is possible to add validation to the query as well, in this case, validating the city name is a real city name before passing it into the database, and rejecting it otherwise.
+
+The sanitization needed for this query is escaping the additional backtick character.
+
+`SanitizedInput = &#96;&#96;) RETURN 1 as a UNION MATCH (n) RETURN 1 WITH DISTINCT true AS haxxored MATCH (n) DETACH DELETE n; \\`
+
+The additional backtick added now ensures that the entire string is used as the node label, and not able to break out of that context.
+
+The unicode character for the bactick; `\u0060` will also resolve to a backtick and needs to be sanitized.
+When handling user input, it is important that the programming language the client is written in is taken into account.
+For example, the input: `\u005C\u00750060` may be resolved before being passed to the database as `\u0060`
+(`\u005C` is backslash, `\`, and `\u0075` is u), which will then be resolved by the database as a backtick!
+
+Writing your own sanitization function can be tricky.
+That is why it is highly recommended to avoid string concatenation and design your database in such a way that user input
+is not needed to dynamically query based on node labels, relationship types and parameters.
+
+It is also important to remember that not all injection attacks rely on quotation escaping of string literals,
+for example the following query, for a number literal, is still vulnerable when not using parameterization.
+
+`query = "MATCH (n) WHERE n.id =" + userid + ";"`
+
+Where the malicious userid input is: `"1 OR 1 = 1 WITH DISTINCT true AS haxxored MATCH (all) DETACH DELETE all; //"`
+
+=== Validation and Sanitization common Exploits
+
+Sanitization can also be used as a technique to clean up user input.
+Another way of keeping the input safe and clean is to use validation.
+Validation checks the input and makes sure it meets a set of certain criteria and will reject the input if it does not,
+in comparison to sanitization which cleans the input only.
+Validation can be used alongside sanitization, keep in mind that both techniques come with risks.
+
+=== Whitespace checks
+
+Checking user input for whitespace sounds like a good way to avoid injection, and in some cases it would work,
+consider the example:
+
+`"Robby' MATCH (s:Student) DETACH DELETE s //"`
+
+A validation check for whitespace would flag this query as invalid, but checking for whitespace alone isn’t enough.
+In Cypher using block comments to replace whitespaces is also valid, the following query would, therefore, pass whitespace validation checks:
+
+`"Robby'/&#42;&#42;/MATCH/&#42;&#42;/(s:Student)/&#42;&#42;/DETACH/&#42;&#42;/DELETE/&#42;&#42;/s/&#42;&#42;///"`
```

Comment:
I'm thinking this is an encoding error when copying from Google doc to .adoc

Reply:
As mentioned earlier, this is intentional to get around asciidocs formatting :) just the html entity encoding of the character!